namesys/publisher.go

```diff
@@ -12,103 +14,196 @@ import (
+func (p *IpnsPublisher) Publish(ctx context.Context, k ci.PrivKey, value path.Path) error {
 	log.Debugf("Publish %s", value)
 	return p.PublishWithEOL(ctx, k, value, time.Now().Add(DefaultRecordTTL))
 }
 
-// PublishWithEOL is a temporary stand in for the ipns records implementation
-// see here for more details: https://github.com/ipfs/specs/tree/master/records
-func (p *ipnsPublisher) PublishWithEOL(ctx context.Context, k ci.PrivKey, value path.Path, eol time.Time) error {
+func IpnsDsKey(id peer.ID) ds.Key {
+	return ds.NewKey("/ipns/" + base32.RawStdEncoding.EncodeToString([]byte(id)))
+}
+
+// PublishedNames returns the latest IPNS records published by this node and
+// their expiration times.
+//
+// This method will not search the routing system for records published by other
+// nodes.
+func (p *IpnsPublisher) ListPublished(ctx context.Context) (map[peer.ID]*pb.IpnsEntry, error) {
+	query, err := p.ds.Query(dsquery.Query{
+		Prefix: ipnsPrefix,
+	})
+	if err != nil {
+		return nil, err
+	}
+	defer query.Close()
+
+	records := make(map[peer.ID]*pb.IpnsEntry)
+	for {
+		select {
+		case result, ok := <-query.Next():
+			if !ok {
+				return records, nil
+			}
+			if result.Error != nil {
+				return nil, result.Error
+			}
+			value, ok := result.Value.([]byte)
+			if !ok {
+				log.Error("found ipns record that we couldn't convert to a value")
+				continue
+			}
+			e := new(pb.IpnsEntry)
+			if err := proto.Unmarshal(value, e); err != nil {
+				// Might as well return what we can.
+				log.Error("found an invalid IPNS entry:", err)
+				continue
+			}
+			if !strings.HasPrefix(result.Key, ipnsPrefix) {
+				log.Errorf("datastore query for keys with prefix %s returned a key: %s", ipnsPrefix, result.Key)
+				continue
+			}
+			k := result.Key[len(ipnsPrefix):]
+			pid, err := base32.RawStdEncoding.DecodeString(k)
+			if err != nil {
+				log.Errorf("ipns ds key invalid: %s", result.Key)
+				continue
+			}
+			records[peer.ID(pid)] = e
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+	}
+}
+
+// GetPublished returns the record this node has published corresponding to the
+// given peer ID.
+//
+// If `checkRouting` is true and we have no existing record, this method will
+// check the routing system for any existing records.
+func (p *IpnsPublisher) GetPublished(ctx context.Context, id peer.ID, checkRouting bool) (*pb.IpnsEntry, error) {
+	ctx, cancel := context.WithTimeout(ctx, time.Second*30)
+	defer cancel()
+
+	dsVal, err := p.ds.Get(IpnsDsKey(id))
+	var value []byte
+	switch err {
+	case nil:
+		var ok bool
+		value, ok = dsVal.([]byte)
+		if !ok {
+			return nil, fmt.Errorf("found ipns record that we couldn't convert to a value")
+		}
+	case ds.ErrNotFound:
+		if !checkRouting {
+			return nil, nil
+		}
+		_, ipnskey := IpnsKeysForID(id)
+		value, err = p.routing.GetValue(ctx, ipnskey)
+		if err != nil {
+			// Not found or other network issue. Can't really do
```

Comment:
dropping errors makes me nervous, would appreciate maybe a debug or info log of this

Reply:
Addressed.